### PR TITLE
[CSL-1854] Get rid of MonadBlock abstraction

### PR DESCRIPTION
**Pull request by pva701**

### CODE REVIEW

File: db/Pos/DB/GState/Common.hs [hunk 1]
```diff
@@ -69,29 +75,25 @@ writeBatchGState = dbWriteBatch' GStateDB
...
-getTipBlockGeneric
-    :: forall block header undo m.
-       MonadBlockDBGeneric header block undo m
-    => m block
-getTipBlockGeneric = getTipSomething "block" (dbGetBlock @_ @block)
+getTipBlock :: MonadDBRead m => m Block
+getTipBlock = getTipSomething "block" dbGetBlock
+
+-- | Returns header of block that was requested from Block DB.
+getHeader
```
Comment (gromakovsky) [inline db/Pos/DB/GState/Common.hs]: Why is it in `GState` module? It's not stored in GState DB, it's stored in Block DB.

File: update/Pos/Update/Poll/Modifier.hs [hunk 2]
```diff
@@ -1,14 +1,64 @@
--- | Functions working with PollModifier.
+-- | Types related to Poll monad.
```
Comment (gromakovsky) [inline update/Pos/Update/Poll/Modifier.hs]: Strange change.

File: db/Pos/DB/Class.hs [hunk 3]
```diff
@@ -91,6 +92,14 @@ class DBIteratorClass i where
 
 type IterType i = (IterKey i, IterValue i)
 
+newtype RawBlock = RawBlock
+    { unRawBlock :: ByteString
+    }
+
+newtype RawUndo = RawUndo
+    { unRawUndo :: ByteString
```
Comment (gromakovsky) [inline db/Pos/DB/Class.hs]: Imo, it'd be better to have `newtype Serialized a = Serialized ByteString`, but for CSL-1854 this solution is enough.

File: db/Pos/DB/Class.hs [hunk 4]
```diff
@@ -104,13 +113,29 @@ class (HasConfiguration, MonadBaseControl IO m, MonadThrow m) => MonadDBRead m w
...
+    dbGetRawBlock = lift . dbGetRawBlock
+    dbGetRawUndo = lift . dbGetRawUndo
+
+
+type MonadBlockDBRead m = (MonadDBRead m, BlockchainHelpers MainBlockchain)
+
+getBlock :: MonadBlockDBRead m => HeaderHash -> m (Maybe Block)
+getBlock x = do
+    mBS <- fmap unRawBlock <$> dbGetRawBlock x
+    pure $ rightToMaybe . decodeFull =<< mBS
```
Comment (gromakovsky) [inline db/Pos/DB/Class.hs]: Why do you discard decode error? `Nothing` means that there is no block. Decode error should trigger `DBMalfored` exception.

File: db/Pos/DB/Class.hs [hunk 5]
```diff
@@ -138,6 +163,8 @@ class MonadDBRead m => MonadDB m where
     -- with given key from DB corresponding to given tag.
     dbDelete :: DBTag -> ByteString -> m ()
 
+    -- | Put given blund into the Block DB.
+    dbPutRawBlund :: (Block, ByteString) -> m ()
```
Comment (gromakovsky) [inline db/Pos/DB/Class.hs]: Why not `RawUndo`?

File: db/Pos/DB/Class.hs [hunk 6]
```diff
@@ -39,6 +39,10 @@ module Pos.DB.Class
        , DBIteratorClass (..)
```
Comment (gromakovsky) [inline db/Pos/DB/Class.hs]: You forget to update the description of this module.

File: lib/src/Pos/Block/Types.hs [hunk 7]
```diff
@@ -33,6 +34,8 @@ instance NFData Undo
 -- | Block and its Undo.
 type Blund = (Block, Undo)
 
+type RawBlund = (Block, ByteString)
```
Comment (gromakovsky) [inline lib/src/Pos/Block/Types.hs]: Why not `RawUndo`?

File: lib/src/Pos/DB/Block.hs [hunk 8]
```diff
@@ -72,27 +47,49 @@ import           System.IO.Error (IOError, isDoesNotExistError)
...
+
+
+----------------------------------------------------------------------------
+-- BlockDB related methods
+----------------------------------------------------------------------------
+
+getUndo :: MonadDBRead m => HeaderHash -> m (Maybe Undo)
+getUndo x = do
+    mBS <- fmap unRawUndo <$> dbGetRawUndo x
+    pure $ rightToMaybe . decodeFull =<< mBS
```
Comment (gromakovsky) [inline lib/src/Pos/DB/Block.hs]: Same question (see `getBlock`).

File: lib/src/Pos/DB/Block.hs [hunk 9]
```diff
@@ -496,11 +273,19 @@ getData fp = handle handler $ liftIO $
...
 
+getRawData ::  forall m . MonadIO m => FilePath -> m ByteString
+getRawData = liftIO . BS.readFile
+
 putData ::  (MonadIO m, Bi v) => FilePath -> v -> m ()
 putData fp v = liftIO $
     bracket (openBinaryFile fp WriteMode) hClose $ \h ->
         BS.hPut h (serialize' v) >> hFlush h
 
+putRawData ::  MonadIO m => FilePath -> ByteString -> m ()
```
Comment (gromakovsky) [inline lib/src/Pos/DB/Block.hs]: Get rid of this copy-paste please.

File: lib/src/Pos/DB/Block.hs [hunk 10]
```diff
@@ -101,151 +98,44 @@ import           Pos.Util.Chrono (NewestFirst (..))
...
--- | Returns header of block that was requested from Block DB.
-blkGetHeader
-    :: (HasConfiguration, MonadDBRead m)
-    => HeaderHash -> m (Maybe BlockHeader)
-blkGetHeader = dbGetBi BlockIndexDB . blockIndexKey
+    => HeaderHash -> m (Maybe ByteString)
+getRawBlock = blockDataPath >=> getData
 
 -- Get undo data for block with given hash from Block DB. This
 -- function has too strict constraint, consider using 'blkGetUndo'.
```
Comment (gromakovsky) [inline lib/src/Pos/DB/Block.hs]: The comment is outdated.

File: lib/src/Pos/DB/Block.hs [hunk 11]
```diff
@@ -101,151 +98,44 @@ import           Pos.Util.Chrono (NewestFirst (..))
...
-getUndo :: (HasConfiguration, MonadRealDB ctx m) => HeaderHash -> m (Maybe Undo)
-getUndo = undoDataPath >=> getData
+getRawUndo :: (HasConfiguration, MonadRealDB ctx m) => HeaderHash -> m (Maybe ByteString)
+getRawUndo = undoDataPath >=> getData
```
Comment (gromakovsky) [inline lib/src/Pos/DB/Block.hs]: The implementation is outdated.

File: lib/src/Pos/DB/Block.hs [hunk 12]
```diff
@@ -289,192 +155,103 @@ decodeOrFailPureDB
...
-    => Blund
+    => RawBlund
     -> m ()
-dbPutBlundPureDefault (blk,undo) = do
+dbPutRawBlundPureDefault (blk, rawUndo) = do
+    let blund :: Blund
+        blund =
+            either (error "Couldn't deserialize undo")
+                   (blk,)
+                   (decodeFull rawUndo)
```
Comment (gromakovsky) [inline lib/src/Pos/DB/Block.hs]: It's absolutely weird to decode and then immediately encode. I suggest that we store blocks and undos separately in pure db. Can you create an issue at least?

File: update/Pos/Update/Poll/Modifier.hs [hunk 13]
```diff
@@ -1,14 +1,55 @@
...
+                                 NumSoftwareVersion, SoftwareVersion, StakeholderId)
+import           Pos.Core.Update (UpId)
+import           Pos.Slotting.Types (SlottingData)
+import           Pos.Update.Poll.Types (BlockVersionState, ConfirmedProposalState, ProposalState)
+import           Pos.Util.Modifier (MapModifier)
+
+-- | PollModifier is used in verification. It represents operation which
+-- one should apply to global state to obtain result of application of
+-- MemPool or blocks which are verified.
+data PollModifier = PollModifier
```
Comment (gromakovsky) [inline update/Pos/Update/Poll/Modifier.hs]: Why did you move it here?

File: auxx/src/Mode.hs [hunk 14]
```diff
@@ -183,29 +180,14 @@ instance {-# OVERLAPPING #-} CanJsonLog AuxxMode where
 instance HasConfiguration => MonadDBRead AuxxMode where
     dbGet = realModeToAuxx ... dbGet
     dbIterSource tag p = hoist (hoist realModeToAuxx) (dbIterSource tag p)
+    dbGetRawBlock = realModeToAuxx ... DB.dbGetRawBlockRealDefault
```
Comment (int-index) [inline auxx/src/Mode.hs]: This should be `dbGetRawBlock = realModeToAuxx ... dbGetRawBlock` instead. Note that we don't want to choose the `Default` implementation here — if another implementation is used in `Real` mode, we won't have to remember to update `auxx` accordingly.

File: auxx/src/Mode.hs [hunk 15]
```diff
@@ -183,29 +180,14 @@ instance {-# OVERLAPPING #-} CanJsonLog AuxxMode where
...
+    dbGetRawUndo = realModeToAuxx ... DB.dbGetRawUndoRealDefault
```
Comment (int-index) [inline auxx/src/Mode.hs]: Same as above, don't commit to a particular impl here.

File: auxx/src/Mode.hs [hunk 16]
```diff
@@ -183,29 +180,14 @@ instance {-# OVERLAPPING #-} CanJsonLog AuxxMode where
...
-    dbGetUndo   = realModeToAuxx ... dbGetUndo @BlockHeader @Block @Undo
-    dbGetHeader = realModeToAuxx ... dbGetHeader @BlockHeader @Block @Undo
-
-instance (HasConfiguration, HasSscConfiguration) =>
-         MonadBlockDBGeneric (Some IsHeader) SscBlock () AuxxMode
-  where
-    dbGetBlock  = realModeToAuxx ... dbGetBlock
-    dbGetUndo   = realModeToAuxx ... dbGetUndo @(Some IsHeader) @SscBlock @()
-    dbGetHeader = realModeToAuxx ... dbGetHeader @(Some IsHeader) @SscBlock @()
+    dbPutRawBlund = realModeToAuxx ... DB.dbPutRawBlundRealDefault
```
Comment (int-index) [inline auxx/src/Mode.hs]: Same as above, don't commit to a particular impl here.

File: db/Pos/DB/Class.hs (hunk 3 above)
Comment (pva701) [reply db/Pos/DB/Class.hs]: I don't think so, I introduced these wrappers to avoid mismatching `Block` and `Undo` ByteStrings

File: lib/src/Pos/DB/Block.hs (hunk 12 above)
Comment (pva701) [reply lib/src/Pos/DB/Block.hs]: >I suggest that we store blocks and undos separately in pure db.

we don't, why do you pay attention to it at all? :thinking: 
I don't think it does matter how we store blund in pure db

File: auxx/src/Mode.hs (hunk 14 above)
Comment (pva701) [reply auxx/src/Mode.hs]: it's already fixed

Comment (pva701) [reply auxx/src/Mode.hs]: can you take away your diss?

File: update/Pos/Update/Poll/Modifier.hs (hunk 13 above)
Comment (pva701) [reply update/Pos/Update/Poll/Modifier.hs]: Because when I was moving it to `core` this datatype shouldn't be moved and I decided it should be logical split with other datatypes. Also, I find it more logical to put it here.

File: auxx/src/Mode.hs (hunk 14 above)
Comment (int-index) [reply auxx/src/Mode.hs]: It's not fixed as far as I can see.

File: lib/src/Pos/DB/Block.hs [hunk 17]
```diff
@@ -72,180 +47,92 @@ import           System.IO.Error (IOError, isDoesNotExistError)
...
--- Implementations for 'MonadRealDB'
+-- BlockDB related methods
 ----------------------------------------------------------------------------
 
--- Get block with given hash from Block DB.  This function has too
--- strict constraint, consider using 'blkGetBlock'.
+getUndo :: MonadDBRead m => HeaderHash -> m (Maybe Undo)
+getUndo x = dbGetRawUndo x >>= \case
+    Nothing      -> pure Nothing
+    Just rawUndo -> eitherToThrow $ first DBMalformed $ decodeFull $ unRawUndo rawUndo
```
Comment (gromakovsky) [inline lib/src/Pos/DB/Block.hs]: Copy-paste 😒

File: lib/src/Pos/DB/Block.hs [hunk 18]
```diff
@@ -72,180 +47,92 @@ import           System.IO.Error (IOError, isDoesNotExistError)
...
+    => HeaderHash -> m (Maybe ByteString)
+getRawBlock = blockDataPath >=> getData
 
--- Get undo data for block with given hash from Block DB. This
--- function has too strict constraint, consider using 'blkGetUndo'.
-getUndo :: (HasConfiguration, MonadRealDB ctx m) => HeaderHash -> m (Maybe Undo)
-getUndo = undoDataPath >=> getData
+-- Get serialization of an undo data for block with given hash from Block DB.
+getRawUndo :: (HasConfiguration, MonadRealDB ctx m) => HeaderHash -> m (Maybe ByteString)
+getRawUndo = undoDataPath >=> getData
```
Comment (gromakovsky) [inline lib/src/Pos/DB/Block.hs]: Once again, how is it supposed to work? `getData` performs `decodeFull`, but for `getRaw*` you don't need to do any decoding.

File: update/Pos/Update/Poll/Modifier.hs (hunk 13 above)
Comment (gromakovsky) [reply update/Pos/Update/Poll/Modifier.hs]: No, we put types into `Types` modules and other stuff into neighbour modules.

**Comment by pva701** [comment]
>No, we put types into Types modules and other stuff into neighbour modules.

I remember, but I don't find at least one reason to do it for this case

File: lib/src/Pos/DB/Block.hs (hunk 12 above)
Comment (gromakovsky) [reply lib/src/Pos/DB/Block.hs]: > we don't

I know that we don't, but I suggest that we do.

> why do you pay attention to it at all?

Because doing unnecessary roundtrip of encoding and decoding is inefficient and it will make tests slower. So I would like to have an issue about it.

**Comment by gromakovsky** [comment]
> I remember, but I don't find at least one reason to do it for this case

Isn't consistency with other code a good reason?

**Comment by pva701** [comment]
no zachem

File: db/Pos/DB/Class.hs [hunk 19]
```diff
@@ -8,6 +8,7 @@
 -- 'MonadDBRead' contains reading and iterating capabilities. The
 -- advantage of it is that you don't need to do any 'IO' to use it
 -- which makes it suitable for pure testing.
+-- 'MonadDBRead' also give you access to the Block DB.
```
Comment (volhovM) [inline db/Pos/DB/Class.hs]: "gives you", but actually "provides"

File: db/Pos/DB/Class.hs (hunk 3 above)
Comment (volhovM) [reply db/Pos/DB/Class.hs]: Pleeease can you _at least_ rename them? We're discussing this CSL-1859 issue and we operate with term `RawBlock` meaning that is an "unverified block", not a "serialized" one. 
I actually find Ivan's idea to be a pretty good thing, because CSL-1869 will anyway introduce a notion of `Serialized a` as Ivan has proposed -- so if you do it now, you minimize the diff.

File: db/Pos/DB/Class.hs [hunk 20]
```diff
@@ -104,13 +106,29 @@ class (HasConfiguration, MonadBaseControl IO m, MonadThrow m) => MonadDBRead m w
...
         MonadDBRead (t m)
   where
     dbGet tag = lift . dbGet tag
     dbIterSource tag (p :: Proxy i) =
         hoist (hoist lift) (dbIterSource tag p)
+    dbGetRawBlock = lift . dbGetRawBlock
+    dbGetRawUndo = lift . dbGetRawUndo
+
+
+type MonadBlockDBRead m = (MonadDBRead m, BlockchainHelpers MainBlockchain)
```
Comment (volhovM) [inline db/Pos/DB/Class.hs]: Can you please point out where do we use `BlockchainHelpers` instance? I don't really see it.

File: db/Pos/DB/Class.hs (hunk 6 above)
Comment (volhovM) [reply db/Pos/DB/Class.hs]: What, do you think, should be added/removed? I will add needed documentation about getting raw data in CSL-1869, so documenting this part is not a strict requirement.

File: db/Pos/DB/GState/Common.hs [hunk 21]
```diff
@@ -6,10 +6,9 @@ module Pos.DB.GState.Common
        (
          -- * Getters
          getTip
+       , getTipSomething
```
Comment (volhovM) [inline db/Pos/DB/GState/Common.hs]: 1. You only use it in a single place.
2. I would prefer `getTipAbstract` (nitpicker-mode-on).

File: db/Pos/DB/BlockIndex.hs [hunk 22]
```diff
@@ -0,0 +1,36 @@
+-- | Operations with block index db.
+
+module Pos.DB.BlockIndex
```
Comment (volhovM) [inline db/Pos/DB/BlockIndex.hs]: I don't get it. Why would you create a module that provides only read capabilities over block index?

AFAIK every other db module follows the following hierarchy:
1. Reader methods
2. Writer batch functions
3. Iterators
4. Prefix keys

It seems I am missing something important. What exactly?

Update: I now seem to have more understanding on this question (after i've read through all the PR), but I would like you to explain anyway. Like, yes -- we always write blunds only, but design is controversial.

File: lib/src/Pos/Block/Types.hs [hunk 23]
```diff
@@ -33,6 +35,8 @@ instance NFData Undo
 -- | Block and its Undo.
 type Blund = (Block, Undo)
 
+type RawBlund = (Block, RawUndo)
```
Comment (volhovM) [inline lib/src/Pos/Block/Types.hs]: Same concern applies (as i addressed for `RawBlock`).

File: lib/src/Pos/Block/Types.hs [hunk 24]
```diff
@@ -47,3 +51,4 @@ instance HasDifficulty Blund where
 
 instance HasHeaderHash Block => HasHeaderHash Blund where
     headerHash = headerHash . fst
+
```
Comment (volhovM) [inline lib/src/Pos/Block/Types.hs]: It's a shame.

File: lib/src/Pos/DB/Block.hs [hunk 25]
```diff
@@ -289,192 +152,103 @@ decodeOrFailPureDB
...
+dbPutRawBlundPureDefault ::
        forall ctx m. (HasConfiguration, MonadPureDB ctx m)
-    => Blund
+    => RawBlund
     -> m ()
-dbPutBlundPureDefault (blk,undo) = do
+dbPutRawBlundPureDefault (blk, rawUndo) = do
+    let blund :: Blund
+        blund =
+            either (error "Couldn't deserialize undo")
```
Comment (volhovM) [inline lib/src/Pos/DB/Block.hs]: It should be `DBMalformed`.

File: lib/src/Pos/Util/BackupPhrase.hs [hunk 26]
```diff
@@ -23,6 +23,7 @@ import           Pos.Crypto (AbstractHash, EncryptedSecretKey, PassPhrase, Secre
                              unsafeAbstractHash)
 import           Pos.Util.Mnemonics (fromMnemonic, toMnemonic)
 import           Test.QuickCheck (Arbitrary (..), genericShrink)
+import           Test.QuickCheck.Instances ()
```
Comment (volhovM) [inline lib/src/Pos/Util/BackupPhrase.hs]: I see no other changes in this file. Why did you decide to import this?

Comment (pva701) [reply lib/src/Pos/Util/BackupPhrase.hs]: because it didn't find `Arbitrary Text` instance when I had merged w/ master. dunno why, gromak give advice to me to add this import and it helped

File: lib/src/Pos/Block/Types.hs (hunk 24 above)
Comment (pva701) [reply lib/src/Pos/Block/Types.hs]: #antihype

File: db/Pos/DB/GState/Common.hs (hunk 21 above)
Comment (pva701) [reply db/Pos/DB/GState/Common.hs]: I use it twice, in this file and in the `lib/DB/Block`

File: db/Pos/DB/Class.hs (hunk 20 above)
Comment (pva701) [reply db/Pos/DB/Class.hs]: when deserialize `Block` to validate validness

File: db/Pos/DB/BlockIndex.hs (hunk 22 above)
Comment (pva701) [reply db/Pos/DB/BlockIndex.hs]: 1. There are only readers methods.
2. We don't write a batch to block index db, just write some information along with putting a block to block db.
3. We don't iterate over headers, we use BlockExtra for it which is stored into GState-db
4. it's done

:paket:

File: db/Pos/DB/Class.hs (hunk 3 above)
Comment (pva701) [reply db/Pos/DB/Class.hs]: I propose `newtype Serialized a = Serialized ByteString`

File: db/Pos/DB/BlockIndex.hs (hunk 22 above)
Comment (volhovM) [reply db/Pos/DB/BlockIndex.hs]: Sorry, but it doesn't really answer my question.
I suggest to export a function instead ofa  prefix key. I find current solution unsafe (everybody from outside can write to `blockIndexKey` something which is different from `HeaderHash`.

File: db/Pos/DB/Class.hs (hunk 3 above)
Comment (pva701) [reply db/Pos/DB/Class.hs]: lol, @gromakovsky it seems you edited your first comment

Comment (volhovM) [reply db/Pos/DB/Class.hs]: The first comment by gromak was there in the same form yesterday evening.
`Serialized a` works for me.

File: db/Pos/DB/BlockIndex.hs (hunk 22 above)
Comment (pva701) [reply db/Pos/DB/BlockIndex.hs]: breten, vso horosho?
`blockIndexKey :: HeaderHash -> ByteString`

**Comment by pva701** [comment]
@volhovM @gromakovsky have been all your major issues resolved? if so, plz approve and write down minor issues you really would like to fix but it can be postponed to next refactoring PR

File: db/Pos/DB/Class.hs (hunk 6 above)
Comment (gromakovsky) [reply db/Pos/DB/Class.hs]: > What, do you think, should be added/removed?

This: https://i.imgur.com/HYITtDr.png

Perhaps those changes were done before you wrote your question, but the thing is they were done after I wrote my comment.

**Comment by pva701** [comment]
@gromakovsky I can, but I really don't think it's important to care about insane behaviour in `PureDB`

**Comment by gromakovsky** [comment]
Tests are slow and I don't think that making them even more slower is a good idea.

**Comment by pva701** [comment]
ah, ok. sound reasonable

**Comment by pva701** [comment]
@volhovM suuda idi

File: db/Pos/DB/BlockIndex.hs (hunk 22 above)
Comment (volhovM) [reply db/Pos/DB/BlockIndex.hs]: It's not an interface we're providing -- that's what I'm trying to say. We don't usually export index keys.
Yeah, whatever, I guess you won't ever understand :pes: :peka:

File: db/Pos/DB/Class.hs (hunk 3 above)
Comment (gromakovsky) [reply db/Pos/DB/Class.hs]: > it seems you've edited your first comment

No, if I edited it, github would show it.
